### PR TITLE
feature: add on-demand vrr

### PR DESCRIPTION
**Pull request by my4ng**
Implements #563.

### CODE REVIEW

File: niri-config/src/lib.rs [hunk 1]
```diff
@@ -323,8 +323,8 @@ pub struct Output {
     pub position: Option<Position>,
     #[knuffel(child, unwrap(argument, str))]
     pub mode: Option<ConfiguredMode>,
-    #[knuffel(child)]
-    pub variable_refresh_rate: bool,
+    #[knuffel(child, default)]
+    pub variable_refresh_rate: Vrr,
```
Comment (YaLTeR) [inline niri-config/src/lib.rs:327]: I think it would be clearer, code- and type-wise, if this worked like FocusFollowsMouse. That is, `pub variable_refresh_rate: Option<Vrr>` where None means disabled, Some means enabled, then on_demand is a bool with default value false.

File: niri-ipc/src/lib.rs [hunk 2]
```diff
@@ -425,6 +418,19 @@ pub struct ConfiguredPosition {
...
+#[cfg_attr(feature = "clap", derive(clap::ValueEnum))]
+#[cfg_attr(feature = "json-schema", derive(schemars::JsonSchema))]
+pub enum VrrToSet {
+    /// Always OFF.
+    AlwaysOff,
+    /// Always ON.
+    AlwaysOn,
+    /// Only ON when a surface matching a VRR window rule is visible on the output.
+    OnDemand,
+}
```
Comment (YaLTeR) [inline niri-ipc/src/lib.rs:432]: I wonder if it's better to match the config here, i.e. instead of `niri msg output vrr always-on/always-off/on-demand` have it be:
- `niri msg output vrr off`
- `niri msg output vrr on`
- `niri msg output vrr on --on-demand`

File: src/backend/mod.rs [hunk 3]
```diff
@@ -153,6 +153,13 @@ impl Backend {
         }
     }
 
+    pub fn set_output_vrr(&mut self, niri: &mut Niri, output: &Output, enable_vrr: bool) {
+        match self {
+            Backend::Tty(tty) => tty.set_output_vrr(niri, output, enable_vrr),
+            Backend::Winit(_) => todo!(),
```
Comment (YaLTeR) [inline src/backend/mod.rs:159]:
```suggestion
            Backend::Winit(_) => (),
```

File: niri-config/src/lib.rs [hunk 4]
```diff
@@ -352,6 +352,32 @@ pub struct Position {
...
 }
 
+#[derive(knuffel::Decode, Debug, Clone, PartialEq, Default)]
+pub struct Vrr {
+    #[knuffel(property, default = Some(false))]
+    on_demand: Option<bool>,
+}
+
+impl Vrr {
+    pub fn always_on(&self) -> bool {
```
Comment (YaLTeR) [inline niri-config/src/lib.rs:362]: The convention is to name checker methods starting from `is_`

File: src/backend/tty.rs [hunk 5]
```diff
@@ -1707,7 +1738,7 @@ impl Tty {
                 };
 
                 let change_mode = surface.compositor.pending_mode() != mode;
-                let change_vrr = surface.vrr_enabled != config.variable_refresh_rate;
+                let change_vrr = surface.vrr_enabled != config.variable_refresh_rate.always_on();
```
Comment (YaLTeR) [inline src/backend/tty.rs:1741]: This seems like it should check whether some surface is enabling VRR currently on this output.

File: src/niri.rs [hunk 6]
```diff
@@ -1200,15 +1201,71 @@ impl State {
...
+                        .invert()
+                        .transform_size(output_size),
+                );
+                let current = workspace.windows().any(|mapped| {
+                    mapped.rules().variable_refresh_rate == Some(true)
+                        && mapped
+                            .window
+                            .geometry()
+                            .to_physical_precise_round(output.current_scale().fractional_scale())
+                            .overlaps(output_geo)
```
Comment (YaLTeR) [inline src/niri.rs:1259]: Window geometry is relative to the window buffer, not a global coordinate system, so this check doesn't work properly.

Either way, I feel like this should check if any surface of the window has this output as its primary output. That will deal with subsurfaces outside the window, and when we get floating, it will also deal with cases like a few pixels of a window overlapping one output (in which case we don't want to enable VRR there). It already works for things like frame callbacks, I think it would be appropriate if the VRR window rule used the same logic.

File: niri-config/src/lib.rs (hunk 1 above)
Comment (my4ng) [reply niri-config/src/lib.rs:327]: That was my initial implementation but the downside is no help function, which is why I wrapped it around a new type.

File: src/backend/mod.rs (hunk 3 above)
Comment (my4ng) [reply src/backend/mod.rs:159]: I blame autocomplete template :p

File: src/backend/tty.rs (hunk 5 above)
Comment (my4ng) [reply src/backend/tty.rs:1741]: Mhmm I am not sure the best way to do that. This just makes it false until the next refresh.

File: niri-ipc/src/lib.rs (hunk 2 above)
Comment (my4ng) [reply niri-ipc/src/lib.rs:432]: I am not sure if this is possible under clap.

File: src/niri.rs (hunk 6 above)
Comment (my4ng) [reply src/niri.rs:1259]: I will have to have a deeper look into the frame callback logic. One issue is that the window may not be in the active workspace and secondly it could be not visible.

File: niri-config/src/lib.rs [hunk 7]
```diff
@@ -324,11 +324,31 @@ pub struct Output {
...
+    pub variable_refresh_rate: Option<Vrr>,
     #[knuffel(child, default = DEFAULT_BACKGROUND_COLOR)]
     pub background_color: Color,
 }
 
+pub trait VrrOutput {
+    fn is_always_on(&self) -> bool;
+    fn is_on_demand(&self) -> bool;
+    fn is_always_off(&self) -> bool;
+}
```
Comment (YaLTeR) [inline niri-config/src/lib.rs:336]: I think it's a better idea to have these methods on `impl Output` directly, and have them contain the word "vrr", i.e. `is_vrr_always_on()`

File: src/backend/tty.rs [hunk 8]
```diff
@@ -1661,6 +1659,39 @@ impl Tty {
...
+                        error!("surface pending connectors is empty");
+                        return;
+                    };
+                    let Some(connector) = device.drm_scanner.connectors().get(&connector) else {
+                        error!("missing enabled connector in drm_scanner");
+                        return;
+                    };
+                    let output_state = niri.output_state.get_mut(output).unwrap();
+
+                    try_to_change_vrr(
```
Comment (YaLTeR) [inline src/backend/tty.rs:1680]: If the surface already has VRR enabled/disabled as we want, then this should just return instead of trying.

File: src/niri.rs [hunk 9]
```diff
@@ -3113,6 +3121,49 @@ impl Niri {
...
+                    let mut visible = false;
+                    mapped.window.with_surfaces(|_, states| {
+                        if !visible {
+                            let surface_primary_scanout_output = states
+                                .data_map
+                                .get_or_insert_threadsafe(Mutex::<PrimaryScanoutOutput>::default);
+                            if surface_primary_scanout_output
+                                .lock()
+                                .unwrap()
+                                .current_output()
```
Comment (YaLTeR) [inline src/niri.rs:3149]: There's a helper for this: `surface_primary_scanout_output(surface, states)`

File: src/niri.rs [hunk 10]
```diff
@@ -3113,6 +3121,49 @@ impl Niri {
...
+            .config
+            .borrow()
+            .outputs
+            .find(&output.name())
+            .map(|output| output.is_on_demand())
+        else {
+            warn!("error getting output config for {}", output.name());
+            return;
+        };
+        if on_demand {
```
Comment (YaLTeR) [inline src/niri.rs:3136]: Would be better to do `if !on_demand { return; }`

File: src/backend/tty.rs [hunk 11]
```diff
@@ -1661,6 +1659,39 @@ impl Tty {
...
+                    let Some(connector) =
+                        surface.compositor.pending_connectors().into_iter().next()
+                    else {
+                        error!("surface pending connectors is empty");
+                        return;
+                    };
+                    let Some(connector) = device.drm_scanner.connectors().get(&connector) else {
+                        error!("missing enabled connector in drm_scanner");
+                        return;
+                    };
```
Comment (YaLTeR) [inline src/backend/tty.rs:1677]: Would it make sense to just store the connector on the `Surface`? I forget already how the code is structured.

File: src/backend/tty.rs (hunk 5 above)
Comment (YaLTeR) [reply src/backend/tty.rs]: Maybe something like, put a bool into OutputState that indicates whether, to the best of our knowledge, VRR should be enabled? Update it in `refresh_on_demand_vrr()`.

Comment (my4ng) [reply src/backend/tty.rs]: The problem is that `refresh_on_demand_vrr` which is called in redraw does not occur before this function, so this won't do any good. I still do not see what the issue is with how it is now, where it just defaults to false. If there is an enabling surface visible, then the display will to VRR upon next redraw anyway.

Comment (YaLTeR) [reply src/backend/tty.rs]: I'm just thinking of a case like changing the output scale or position while there's a surface forcing VRR. Currently this will also toggle the VRR KMS property off and on. Not a big deal I guess, but also when this happens, the output state is not erased, so if you put the last frame's VRR flag there, it should work?

Comment (my4ng) [reply src/backend/tty.rs]: Uh yes I forgot about that case. I was thinking when switching from always on to on demand. I will add a flag there in the output state.

Comment (my4ng) [reply src/backend/tty.rs]: I don't think a simple VRR flag is enough. The window rule config could have changed at the same time, and so what enabled the VRR previously may not apply anymore. There are also edge cases with the other transitions. Overall, I don't think it is worth the trouble to implement it.

**Comment by YaLTeR** [comment]
Did some minor changes + wiki. After a quick test, this appears to work!

**Review by YaLTeR — APPROVED** [review]

**Comment by YaLTeR** [comment]
Thanks!
